Design an intentional way to invalidate platform font caches

litherum opened the pull request:
#### 78afcc43874cb89741751616c94819a3e5e4cc68
<pre>
Design an intentional way to invalidate platform font caches
<a href="https://bugs.webkit.org/show_bug.cgi?id=241537">https://bugs.webkit.org/show_bug.cgi?id=241537</a>

Reviewed by Cameron McCormack and Simon Fraser.

Our platform font invalidation code is fairly spotty - the choice of &quot;what invalidates what&quot; isn&apos;t
particularly deliberate. This patch reorganizes our platform font invalidation code to have this
design (where the arrows represent &quot;A calls B&quot; relationships):

FontCache::invalidateAllFontCaches()
              ||
              ||
              ||
              V
    FontCache::invalidate()
              ||           \\
              ||            \\
              ||              =====
              ||                   \\
              ||                    V
              ||                    SystemFontDatabase::invalidate()
              V                                   ||
FontCache::platformInvalidate()                   ||
                                                  ||
                                                  ||
                                                  V
                                SystemFontDatabase::platformInvalidate()

So now, there is one clear place that should be called whenever this stuff needs to be invalidated:
FontCache::invalidateAllFontCaches(). This greatly simplifies the calculus of &quot;which invalidation function
do I need to run in which situations.&quot; Also, because these invalidation functions are only called during
situations like the memory pressure handler, or when system preferences change, performance isn&apos;t a concern
here, and it&apos;s way more important that the relevant invalidation actually happens than it is to
exactly minimize the set of specific things that need to be invalidated in every specific situation.

On Cocoa ports, SystemFontDatabase::platformInvalidate() is implemented by SystemFontDatabaseCoreText,
which already exists. Also, this patch updates the existing platform notification handlers to call
the top level entry point FontCache::invalidateAllFontCaches() as expected.

This is a defensive patch. There&apos;s no specific bug or change in behavior I&apos;m aiming for; instead I&apos;m just
trying to make it more likely that the right thing happens at the right time. I&apos;m also trying to create
an infrastructure I can plug into when I implement <a href="https://bugs.webkit.org/show_bug.cgi?id=237817.">https://bugs.webkit.org/show_bug.cgi?id=237817.</a>

No tests because this is a defensive patch.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::firstTimeInitialization):
* Source/WebCore/page/cocoa/MemoryReleaseCocoa.mm:
(WebCore::platformReleaseMemory):
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::invalidate):
* Source/WebCore/platform/graphics/FontCache.h:
(WebCore::FontCache::generation const):
* Source/WebCore/platform/graphics/SystemFontDatabase.cpp:
(WebCore::SystemFontDatabase::invalidate):
(WebCore::SystemFontDatabase::clear): Deleted.
* Source/WebCore/platform/graphics/SystemFontDatabase.h:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::fontCacheRegisteredFontsChangedNotificationCallback):
(WebCore::FontCache::platformInvalidate):
(WebCore::invalidateFontCache): Deleted.
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
(WebCore::SystemFontDatabaseCoreText::clear):
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.h:
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::FontCache::platformInvalidate):
* Source/WebCore/platform/graphics/gtk/SystemFontDatabaseGTK.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebCore/platform/graphics/playstation/SystemFontDatabasePlayStation.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::platformInvalidate):
* Source/WebCore/platform/graphics/win/SystemFontDatabaseWin.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebCore/platform/graphics/wpe/SystemFontDatabaseWPE.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::contentSizeCategoryDidChange):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::terminate):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::accessibilityPreferencesDidChange):

Canonical link: <a href="https://commits.webkit.org/251845@main">https://commits.webkit.org/251845@main</a>
</pre>

heycam commented on Source/WebCore/platform/graphics/FontCache.cpp:
```diff
@@ -500,14 +505,14 @@ void FontCache::registerFontCacheInvalidationCallback(Function<void()>&& callbac
 
-    if (fontCacheInvalidationCallback())
+    if (shouldRunInvalidationCallback == ShouldRunInvalidationCallback::Yes && fontCacheInvalidationCallback())
         fontCacheInvalidationCallback()();
```
I wonder if we should replace the callback and its registration mechanism with a direct call to `Page::updateStyleForAllPagesAfterGlobalChangeInEnvironment`. These are both things in WebCore, and I don't think it's a big layering violation.

litherum replied:
platform/ isn't allowed to see outside of platform/. PAL was invented to enforce this layering; it's intended to be a destination for all code that is currently in platform/.

heycam commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -893,7 +893,7 @@ void WebProcess::terminate()
     // These are done in an attempt to reduce LEAK output.
     GCController::singleton().garbageCollectNow();
-    FontCache::forCurrentThread().invalidate();
+    FontCache::invalidateAllFontCaches();
```
This is in `WebProcess::terminate` which by the end of the function will terminate the process. Eventually, `FontCache::invalidateAllFontCaches` will iterate over workers and send them messages to invalidate their thread-specific `FontCache`s. Is it possible workers still exist by the time we get here? (Maybe not?)

smfr commented on Source/WebCore/platform/graphics/FontCache.h:
```diff
@@ -327,9 +327,19 @@ class FontCache {
+    // On the other hand, if we're invalidating because the set of installed fonts changed,
+    // or if some accessibility text settings were altered, we should run a style recalc
+    // so the user can immediately see the effect of the new environment.
+    enum class ShouldRunInvalidationCallback : uint8_t {
```
`: bool`. Some folks put No first since it's the falsey value.

smfr commented on Source/WebCore/platform/graphics/SystemFontDatabase.h:
```diff
@@ -74,12 +74,16 @@ class SystemFontDatabase {
 private:
+    friend class FontCache;
+
+    // Don't call these. Instead, you should be calling FontCache::invalidateAllFontCaches().
```
Maybe them being `private` makes the comment unnecessary?

litherum committed:
Design an intentional way to invalidate platform font caches
https://bugs.webkit.org/show_bug.cgi?id=241537

Reviewed by Cameron McCormack and Simon Fraser.

Our platform font invalidation code is fairly spotty - the choice of "what invalidates what" isn't
particularly deliberate. This patch reorganizes our platform font invalidation code to have this
design (where the arrows represent "A calls B" relationships):

FontCache::invalidateAllFontCaches()
              ||
              ||
              ||
              V
    FontCache::invalidate()
              ||           \\
              ||            \\
              ||              =====
              ||                   \\
              ||                    V
              ||                    SystemFontDatabase::invalidate()
              V                                   ||
FontCache::platformInvalidate()                   ||
                                                  ||
                                                  ||
                                                  V
                                SystemFontDatabase::platformInvalidate()

So now, there is one clear place that should be called whenever this stuff needs to be invalidated:
FontCache::invalidateAllFontCaches(). This greatly simplifies the calculus of "which invalidation function
do I need to run in which situations." Also, because these invalidation functions are only called during
situations like the memory pressure handler, or when system preferences change, performance isn't a concern
here, and it's way more important that the relevant invalidation actually happens than it is to
exactly minimize the set of specific things that need to be invalidated in every specific situation.

On Cocoa ports, SystemFontDatabase::platformInvalidate() is implemented by SystemFontDatabaseCoreText,
which already exists. Also, this patch updates the existing platform notification handlers to call
the top level entry point FontCache::invalidateAllFontCaches() as expected.

This is a defensive patch. There's no specific bug or change in behavior I'm aiming for; instead I'm just
trying to make it more likely that the right thing happens at the right time. I'm also trying to create
an infrastructure I can plug into when I implement https://bugs.webkit.org/show_bug.cgi?id=237817.

No tests because this is a defensive patch.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::firstTimeInitialization):
* Source/WebCore/page/cocoa/MemoryReleaseCocoa.mm:
(WebCore::platformReleaseMemory):
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::invalidate):
* Source/WebCore/platform/graphics/FontCache.h:
(WebCore::FontCache::generation const):
* Source/WebCore/platform/graphics/SystemFontDatabase.cpp:
(WebCore::SystemFontDatabase::invalidate):
(WebCore::SystemFontDatabase::clear): Deleted.
* Source/WebCore/platform/graphics/SystemFontDatabase.h:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::fontCacheRegisteredFontsChangedNotificationCallback):
(WebCore::FontCache::platformInvalidate):
(WebCore::invalidateFontCache): Deleted.
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
(WebCore::SystemFontDatabaseCoreText::clear):
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.h:
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::FontCache::platformInvalidate):
* Source/WebCore/platform/graphics/gtk/SystemFontDatabaseGTK.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebCore/platform/graphics/playstation/SystemFontDatabasePlayStation.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::platformInvalidate):
* Source/WebCore/platform/graphics/win/SystemFontDatabaseWin.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebCore/platform/graphics/wpe/SystemFontDatabaseWPE.cpp:
(WebCore::SystemFontDatabase::platformInvalidate):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::contentSizeCategoryDidChange):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::terminate):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::accessibilityPreferencesDidChange):

Canonical link: https://commits.webkit.org/251845@main
